crates/uv-installer/src/compile.rs

```diff
@@ -0,0 +1,207 @@
+use std::io;
+use std::path::{Path, PathBuf};
+use std::process::Stdio;
+use std::time::Duration;
+
+use async_channel::{Receiver, SendError};
+use tempfile::tempdir;
+use thiserror::Error;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{ChildStdin, ChildStdout, Command};
+use tokio::task::JoinError;
+use walkdir::WalkDir;
+
+const COMPILEALL_SCRIPT: &str = include_str!("pip_compileall.py");
+const MAIN_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Error)]
+pub enum CompileError {
+    #[error("Failed to list files in site-packages")]
+    Walkdir(#[from] walkdir::Error),
+    #[error("Couldn't determine number of cores")]
+    AvailableParallelism(#[source] io::Error),
+    #[error("Failed to send task to worker")]
+    WorkerDisappeared(SendError<PathBuf>),
+    #[error("The task executor is broken, did some other task panic?")]
+    Join(#[from] JoinError),
+    #[error("Failed to start Python interpreter to run compile script")]
+    PythonSubcommand(#[source] io::Error),
+    #[error("Failed to create temporary script file")]
+    TempFile(#[source] io::Error),
+    #[error("Bytecode compilation failed, expected {0:?}, got: {1:?}")]
+    WrongPath(String, String),
+    #[error("Failed to write to Python stdin")]
+    ChildStdin(#[source] io::Error),
+    #[error("Failed to read Python stdout")]
+    ChildStdout(#[source] io::Error),
+    #[error("Bytecode timed out ({}s)", _0.as_secs_f32())]
+    Timeout(Duration),
+}
+
+/// Bytecode compile all file in `dir` using a pool of work-stealing python interpreters running a
+/// Python script that calls `compileall.compile_file`.
+///
+/// All compilation errors are muted (like pip). There is a 10s timeout to catch anything gotten
+/// stuck; This happens way to easily with channels and subprocesses, e.g. because some pipe is
+/// full, we're waiting when it's buffered or we didn't handle channel closing properly.
+///
+/// We only compile all files, but we don't update the RECORD, relying on PEP 491:
+/// > Uninstallers should be smart enough to remove .pyc even if it is not mentioned in RECORD.
+pub async fn compile_tree(dir: &Path, python_executable: &Path) -> Result<usize, CompileError> {
+    let workers =
+        std::thread::available_parallelism().map_err(CompileError::AvailableParallelism)?;
+    // 10 is an arbitrary number.
+    let (sender, receiver) = async_channel::bounded::<PathBuf>(workers.get() * 10);
```

Comment:
I wouldn't even bother with `* 10`.

You're probably find with a rendezvous channel (a capacity of `0`). Rendezvous channels are especially nice because they manifest logic errors in the structure of your concurrency very quickly. (Although I do think there are occasions where using a capacity of `1` can lead to simplifications that you couldn't otherwise do with a capacity of `0`.)

It's like how most times, we just use `vec![]` because you usually don't benefit from starting with some non-zero initial capacity.

Reply:
A larger queue is significantly faster. I tried size 1, the number of workers, the number of workers by 10 and size 1000.

```
hyperfine --warmup 1 --prepare "target/profiling/uv-dev clear-compile .venv" "target/profiling/uv-dev1 compile .venv" "target/profiling/uv-devworkers compile .venv" "target/profiling/uv-devworkers10 compile .venv" "target/profiling/uv-dev1000 compile .venv"
Benchmark 1: target/profiling/uv-dev1 compile .venv
  Time (mean ± σ):     359.8 ms ±  11.5 ms    [User: 6757.8 ms, System: 1849.4 ms]
  Range (min … max):   346.8 ms … 380.1 ms    10 runs
 
Benchmark 2: target/profiling/uv-devworkers compile .venv
  Time (mean ± σ):     352.6 ms ±   8.1 ms    [User: 6856.0 ms, System: 1709.1 ms]
  Range (min … max):   337.5 ms … 368.1 ms    10 runs
 
Benchmark 3: target/profiling/uv-devworkers10 compile .venv
  Time (mean ± σ):     331.4 ms ±   5.9 ms    [User: 6723.6 ms, System: 1761.8 ms]
  Range (min … max):   323.9 ms … 340.1 ms    10 runs
 
Benchmark 4: target/profiling/uv-dev1000 compile .venv
  Time (mean ± σ):     331.0 ms ±   6.0 ms    [User: 6735.2 ms, System: 1738.5 ms]
  Range (min … max):   323.7 ms … 339.6 ms    10 runs
 
Summary
  target/profiling/uv-dev1000 compile .venv ran
    1.00 ± 0.03 times faster than target/profiling/uv-devworkers10 compile .venv
    1.07 ± 0.03 times faster than target/profiling/uv-devworkers compile .venv
    1.09 ± 0.04 times faster than target/profiling/uv-dev1 compile .venv
```

Reply:
That is quite interesting. I stand corrected. It might be good to add a comment about it here (no need to include the full benchmark results though).

To me I think this implies that the directory traversal isn't feeding file paths fast enough into the workers, right? And I think that would be exacerbated with a high core count. It means there are likely idle workers in the low capacity case where as those workers get work more quickly in the high capacity case. Very interesting.